### PR TITLE
Add JMH support and improve MultiSpanExporter performance

### PR DESCRIPTION
**Pull request by sfriberg**
This PR adds support for building benchmarks as part of opentelemetry using JMH. Simply add `benchmarks/java/` folder under source and your benchmark and an artifact with `subproject-benchmark.jar` will be built with them. To run either use `gradle benchmarkRun` or `java -jar XYZ-benchmark.jar` if you have copied/downloaded all artifacts to the same directory.  

I added a single microbenchmark and did an optimization to the MultiSpanExporter implementation to avoid allocation that happens when you are doing iteration of a unmodifiable collection, each entry returned by the iterator will be wrapped by an unmodifiable entry to ensure no modifications can be done.

```
Running on my laptop so not the most stable env...

Before
Benchmark                          (exporterCount)  (spanCount)   Mode  Cnt      Score      Error   Units
MultiSpanExporterBenchmark.export                1         1000  thrpt    5  85884.912 ± 3031.709  ops/ms
MultiSpanExporterBenchmark.export                3         1000  thrpt    5  68166.748 ±  985.945  ops/ms

After
Benchmark                          (exporterCount)  (spanCount)   Mode  Cnt       Score       Error   Units
MultiSpanExporterBenchmark.export                1         1000  thrpt    5  213964.678 ± 62705.536  ops/ms
MultiSpanExporterBenchmark.export                3         1000  thrpt    5  163793.449 ± 53344.890  ops/ms
```

If someone has a good idea on how to avoid the ErrorProne warnings when compiling the generated JMH code please let me know.

### CODE REVIEW

File: benchmark.gradle [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	options.compilerArgs = compileJava.options.compilerArgs
+	options.fork = compileJava.options.fork
+	options.forkOptions = compileJava.options.forkOptions
+}
+
+task benchmarkRun(type: JavaExec) {
+	enabled benchmarksAvailable
+	description 'Run benchmarks. Use property benchmarkFilter to set benchmark filtering regex.'
+	group 'Verification'
+	classpath = sourceSets.benchmark.runtimeClasspath
```
Comment (bogdandrutu) [inline benchmark.gradle]: I think jmh works by default with jmh directory not benchmark, am I wrong?

**Comment by bogdandrutu** [comment]
@sfriberg can you please sign the CLA and respond to the comment(s)?

File: benchmark.gradle (hunk 1 above)
Comment (sfriberg) [reply benchmark.gradle]: Not sure there is a 'default' directory. I opted for benchmark in similar fashion to 'test' which describes the purpose of the code in the sourceset, but not the test framework (junit, testng, ...).

No problem to change if there is a preference for jmh.

Working on getting the CLA signed.

**Comment by jkwatson** [comment]
@sfriberg any update on the CLA?

**Comment by sfriberg** [comment]
Got some issues with my CNCF account so have emailed them to see if it can be resolved so I can sign the CLA

**Comment by sfriberg** [comment]
I signed it

File: benchmark.gradle (hunk 1 above)
Comment (jkwatson) [reply benchmark.gradle]: If `jmh` is the default, then that's the way I would go, as well.

**Comment by jkwatson** [comment]
Before we merge this, can you rebase off of master and try running the benchmark? I think the SpanData creation in here will probably fail due to some changes in that class between when you wrote it and now.

**Comment by sfriberg** [comment]
@jkwatson Thanks, yes had to do a minor change to make it work

Added some comments in the script about how to use it as well so I think it should be in a good state to merge now.

**Comment by jkwatson** [comment]
:shipit:

**Pull request merged**